src/ResourceManager/Common/Commands.ResourceManager.Common/ArgumentCompleters/ScopeCompleter.cs

```diff
@@ -0,0 +1,139 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// ----------------------------------------------------------------------------------
+
+namespace Microsoft.Azure.Commands.ResourceManager.Common.ArgumentCompleters
+{
+    using Commands.Common.Authentication;
+    using Commands.Common.Authentication.Abstractions;
+    using Management.Internal.Resources;
+    using Management.Internal.Resources.Models;
+    using Properties;
+    using Rest.Azure;
+    using System;
+    using System.Collections.Concurrent;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Management.Automation;
+
+    public class ScopeCompleterAttribute : ArgumentCompleterAttribute
+    {
+        public static IDictionary<int, IList<string>> _scopeDictionary = new ConcurrentDictionary<int, IList<string>>();
+        private static readonly object _lock = new object();
+        public static int _timeout = 3;
+
+        protected static IList<String> Scopes
+        {
+            get
+            {
+                lock (_lock)
+                {
+                    IAzureContext context = AzureRmProfileProvider.Instance.Profile.DefaultContext;
+                    var contextHash = HashContext(context);
+                    if (!_scopeDictionary.ContainsKey(contextHash))
+                    {
+                        var tempScopeList = new List<string>();
+                        try
+                        {
+                            var instance = AzureSession.Instance;
+                            var client = instance.ClientFactory.CreateCustomArmClient<ResourceManagementClient>(
+                                context.Environment.GetEndpointAsUri(AzureEnvironment.Endpoint.ResourceManager),
+                                instance.AuthenticationFactory.GetServiceClientCredentials(context, AzureEnvironment.Endpoint.ResourceManager),
+                                instance.ClientFactory.GetCustomHandlers());
+                            client.SubscriptionId = context.Subscription.Id;
+                            // Retrieve only the first page of ResourceGroups to display to the user
+                            var resourceGroups = client.ResourceGroups.ListAsync();
+                            if (resourceGroups.Wait(TimeSpan.FromSeconds(_timeout)))
+                            {
+                                tempScopeList = CreateScopeList(resourceGroups.Result, context.Subscription.Id);
+                                _scopeDictionary[contextHash] = tempScopeList;
+                            }
+#if DEBUG
+                            else
+                            {
+                                throw new InvalidOperationException("client.ResourceGroups call timed out");
+                            }
+#endif
+                        }
+
+                        catch (Exception ex)
+                        {
+                            if (ex == null) { }
```

Comment:
I think this check should also be in the ifdef

Reply:
@markcowl this was modeled after the `ResourceGroupCompleter`, but I believe that it's here so that the release build won't fail due to an unused `ex` variable